mds: add truncate size handling support for fscrypt

lxbsz opened the pull request:
This PR is to support the size handling(truncate size) feature for file cryption, and this is based on Jeff's https://github.com/ceph/ceph/pull/41284 PR.

In the kclient side, only when truncating smaller size and the new size is not aligned to `CEPH_FSCRYPT_BLOCK_SIZE` will the truncate request pass the encrypted last block contents to MDS. And the MDS will help update the last block when truncating the file size.

More detail please see the commit comments.



<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

jtlayton commented on src/mds/MDCache.cc:
```diff
@@ -6469,13 +6471,54 @@ void MDCache::_truncate_inode(CInode *in, LogSegment *ls)
   }
-  dout(10) << "_truncate_inode  snapc " << snapc << " on " << *in << dendl;
+  dout(10) << "_truncate_inode  snapc " << snapc << " on " << *in
+	   << " fscrypt_last_block lengh is " << pi->fscrypt_last_block.length()
```
nit: "fscrypt_last_block length"

lxbsz replied:
Will fix it.

jtlayton commented on src/mds/mdstypes.h:
```diff
@@ -627,6 +631,8 @@ struct inode_t {
   std::vector<uint8_t> fscrypt_file;
 
+  bufferlist fscrypt_last_block;
+
```
Hmm, so this lives in the inode_t? I forget -- is inode_t a long-lived structure in the MDS that describes an inode or just a transient description of the inode that lives in a per-call fashion? If it's the former, then how is this value protected against racing truncate calls from different clients?

lxbsz replied:
It seems is, I will fix this and move this to a proper place.

lxbsz replied:
Currently the last block contents will be journaled together with the project inode and it will be cleared after the truncate finishes, and won't make it persistent together with the `CInode:inode_t` in the disks.

jtlayton commented on src/mds/journal.cc:
```diff
@@ -520,6 +521,9 @@ void EMetaBlob::fullbit::decode(bufferlist::const_iterator &bl) {
     decode(alternate_name, bl);
   }
+  if (struct_v >= 10) {
+    decode(fscrypt_last_block, bl);
```
I'm not sure if this is the place to do it, but we probably do want to add a mds_fscrypt_last_block_max_size tunable or something so that the client can't send us truncate requests with an enormous fscrypt_last_block. We could probably make the default 4k, but since it's a max, you could make it larger (8k or so).

lxbsz replied:
Okay, will add that.

lxbsz commented:
@kotreshhr Could help review the `mds: add truncate size handling support for fscrypt` patch in this PR, maybe we need to fix the journal log related code as Jeff mentioned above.

Thanks.

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -6469,13 +6451,59 @@ void MDCache::_truncate_inode(CInode *in, LogSegment *ls)
+    dout(10) << "_truncate_inode write_trunc on inode " << *in << " objver: "
+             << header.objver << " offset: " << header.file_offset << " blen: "
+	     << header.block_size << dendl;
+    filer.write_trunc(in->ino(), &layout, *snapc, header.file_offset, header.block_size,
```
The `write_trunc()` works well in case the `truncate_from - truncate_size < 4MB` case, in other case it will keep the stale contents after truncate to a smaller size then a bigger one.

jtlayton replied:
Nice catch! I'll plan to do some testing with this as well.

lxbsz commented:
Have removed the `mds: bump truncate seq when fscrypt_file changes` commit as discussed in https://github.com/ceph/ceph/pull/41284#discussion_r750812786.

lxbsz commented on src/mds/CInode.cc:
```diff
@@ -4139,6 +4139,7 @@ void CInode::encode_cap_message(const ref_t<MClientCaps> &m, Capability *cap)
   m->size = i->size;
   m->truncate_seq = i->truncate_seq;
   m->truncate_size = i->truncate_size;
+  m->fscrypt_file = pi->fscrypt_file;
```
For the `fscrypt_auth`, for now I don't will it be used in client side when updating the caps.

Will it make sense to send it to kclient here in future uses ?

lxbsz commented:
Cleaned up patches and added more comments, and also split the big patch into several small ones. Since Jeff and I have passed all the xfstests tool related tests in kclient with this PR, I have dropped the Draft flag.

lxbsz commented:
jenkins test make check arm64

jtlayton commented on src/mds/Server.cc:
```diff
@@ -5089,6 +5126,64 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
+
+      if (!header.objver) {
+        dout(20) << __func__ << " The last block is located in a file hole!"
+		 << dendl;
```
Well, it was in a hole when the client did the read, but it might not be a hole anymore by this point in time. I think if you get a 0 objver from the client, then the operation needs to be gated on whether the backing object exists when you go to do the truncate.

lxbsz replied:
So strange, in many PRs I couldn't receive any notification about the new comments for recent weeks.

Yeah, make sense and I will fix it.

jtlayton commented on src/mds/Server.cc:
```diff
@@ -5089,6 +5126,64 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
+        dout(20) << __func__ << " The last block is located in a file hole!"
+		 << dendl;
+      } else if (mdr->objver) {
+        if (mdr->objver != header.objver) {
```
This is not at all what I had in mind. I had assumed that the objver you got from the client would be used to build an assert_ver op that preceded the truncate and write of the backing object. Here though, you're comparing the objver in the setattr request to mdr->objver.

Where does mdr->objver come from and how do you know that it's the same value as the backing object of the last block? I guess you're relying on Fx ensuring that no changes can be made? That may be OK then.

lxbsz replied:
The `mdr->objver`, which initially is 0, is read from the Rados in Line#5183 below. Yeah, since the setattr has `xlock(filelock)` at the beginning, so no client could write to the file during this time.

We need to make sure the `header.objver` will be the latest after xlocked the filelock as you mentioned above.

lxbsz replied:
We need to verify the object versions before submitting the mdlog for it, or it will be more complicated if we need to let client retry if they won't match.

jtlayton replied:
Ok, I suppose we can live with that. It's an extra round trip between the MDS and OSD, but that's not the end of the world. I'll mark this resolved.

jtlayton commented on src/mds/Server.cc:
```diff
@@ -5061,15 +5115,86 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
+                                  new C_OnFinisher(new C_IO_MDC_ReadtruncFinish(mdcache, cur, mdr, data),
+                                                   mds->finisher),
+                                  &mdr->objver);
+        return;
```
This still doesn't look right to me, as I don't see how you're handling the case where the last block is inside an existing hole in the file.

IOW, suppose the client reads the last block for a truncate and and finds that the backing object does not exist. The client sends a 0 object version in the setattr request, along with an encrypted last block. However, before that setattr request arrives at the MDS another client issues a write and creates the object and writes some data to it.

What happens at that point? I don't see where you're ensuring that the trailing object still no longer exists when you go to issue the truncate and write.

lxbsz replied:
> This still doesn't look right to me, as I don't see how you're handling the case where the last block is inside an existing hole in the file.
> 
> IOW, suppose the client reads the last block for a truncate and and finds that the backing object does not exist. The client sends a 0 object version in the setattr request, along with an encrypted last block. However, before that setattr request arrives at the MDS another client issues a write and creates the object and writes some data to it.
> 
> What happens at that point? I don't see where you're ensuring that the trailing object still no longer exists when you go to issue the truncate and write.

In Line#5163 above will ensure this. In that case the objvers won't be same and kclient will set the `header.objver = 0`, and in MDS the new read `mdr->objver > 0`. Will let kclient retry it.

NOTE: The `mdr->objver` will be initialized to `ULONG_MAX` and will always read that object to get the `objver` just after xlock succeeds.

The logic here is:

1, kclient will read that object and if it's a hole, then just set the `header.objver = 0`.
2, kclient sends the setattr request to MDS.
3, in MDS it will try to `xlock` the `in->filelock` first, which will revoke the `Fsxrw` caps from all the clients, so no client could write new contents to the file.
4, in MDS it will try to read that object just after `xlock` succeeds, if the new read `objver` is changed. No matter it becomes a hole or written new contents to it. The MDS will fail the request and let kclient to retry it.
5, if the `objver` doesn't change since Step1, then we can continue.
6, journal the request, early reply the request
7, flush the journal
8, apply the change
9, drop the `xlock(in->filelock)` and safe reply the request

lxbsz commented:
There has another corner case we need to handle:

1, kclient reads the object and the `header.objver = 1`.
2, kclient sends the setattr request to MDS.
3, in MDS it will try to xlock the `in->filelock` first, which will revoke the Fsxrw caps from all the clients, so no client could write new contents to the file.
4, in MDS it will try to read that object just after xlocking succeeds, the new read `objver` is 1 too.
5, if the objver doesn't change since Step1, then we can continue.
6, ...

In Step4 above, before the MDS successfully acquired the `xlock(in->filelock)`, will it possible that another client could delete that object and write new contents to it again, and set the `objver` to 1 too ?

Then we need to retry the request anyway !

I think there have two ways to resolve this:
1, in kclient get the file's `mtime` and sends it together with the `header.objver` to MDS in the request.
2, in MDS after xlocking the `in->filelock` check whether the `mtime` changed or not.
3, if the `mtime` changed, retry it directly or will check the `objvers`...

Another way is if we can get the object's creating or modified timestamp in the MOSDReply together with the `objver` ? Then we can check the object's `timestamps` and `objvers` at the same time. 

Any idea ?

jtlayton commented:
It's a somewhat unlikely race, but it is possible. The problem with using mtime is that it only has granularity of jiffies. With modern computers a lot can happen before the mtime can change.
It's generally better to work with i_version for this reason. 

So, we could switch all of this to use the i_version of the inode instead of the underlying object version to gate the operation. You're already relying on caps to prevent things from changing, so that should be just as good. We'll need to change the kernel patches, but it should be doable.

lxbsz commented:
> It's a somewhat unlikely race, but it is possible. The problem with using mtime is that it only has granularity of jiffies. With modern computers a lot can happen before the mtime can change. It's generally better to work with i_version for this reason.
> 
> So, we could switch all of this to use the i_version of the inode instead of the underlying object version to gate the operation. You're already relying on caps to prevent things from changing, so that should be just as good. We'll need to change the kernel patches, but it should be doable.
> 
> I probably wouldn't bother with both the object and inode version for this. Just the inode version should be fine.

If we only use the `i_version` here, I am afraid it will be racy in case likes:

ClientA: in a loop keep touching the `fileX`, which will only change the access and modification times.
ClinetB: try to truncate the file.

For the ClientB it's possibly will keep retrying since the ClientA will project the CInode and increase the `i_version` ?

Or some other similar cases like the above.

IMO, the `objver` has granularity of Rados objects, we could continue to use the `objver` just like what it is now and only when the `objvers` are different will comparing the `i_version` is needed. Because another client deletes and then creates the same Rados object during we are truncating a file is very unlikely to happen, and should be in very rare case. So mostly the `objver` should be enough.

Any idea ?

jtlayton commented:
I think we shouldn't worry too much about livelocking like that. There isn't really much we can do about it. Eventually the kernel will just return -EAGAIN or whatever in this situation.

We may just have to mention that in the release notes and hope that it doesn't happen much. If it becomes a big problem in the future we can discuss other ways to avoid it. One idea: when the MDS is going to return -EAGAIN on the truncate, it could revoke Fw from everyone (or whatever cap governs mtime), and then hold off on handing it out again for a few seconds (or until the truncate is successful).

OTOH, I'd not implement this just yet, until we have an idea of how big a problem this is in practice.

lxbsz commented:
> I think we shouldn't worry too much about livelocking like that. There isn't really much we can do about it. Eventually the kernel will just return -EAGAIN or whatever in this situation.
> 
Yeah, agree. Maybe we can increase the retry count in kclient ?

> We may just have to mention that in the release notes and hope that it doesn't happen much. If it becomes a big problem in the future we can discuss other ways to avoid it. One idea: when the MDS is going to return -EAGAIN on the truncate, it could revoke Fw from everyone (or whatever cap governs mtime), and then hold off on handing it out again for a few seconds (or until the truncate is successful).

Not only the mtime will increase the `i_version` but also others. So let's just skip this rare corner case for now and add some comments on it.

> 
> OTOH, I'd not implement this just yet, until we have an idea of how big a problem this is in practice.

Sure, I will switch `objver` to `i_version` tomorrow in both kclient and this PR.

lxbsz commented:
@jtlayton 

Today I tried the `i_version` but it won't work, that's because:

1, in the kclient before sending the setattr request to MDS I have pulled the lasted `i_version` from MDS side by using both `getattr` and try to update the  `i_version` in each cap grant/update. Please note at the same time we need to get the `Fr` cap to read the last block from Rados in this step.
2, in Step1 in most cases we can be sure that the `i_version` will be the newest, then we can send it to MDS in setattr request.
3, in MDS when handling the setattr it will xlockes the `in->filelock`, which will try to revoke the `Fr` caps.
4, when revoking the `Fr` caps, the `Locker.cc` will project the inode and increase the `i_version`.
5, so once the MDS successfully xlockes the `in->filelock`, the `i_version` will always be updated.

That means in MDS when comparing the `i_versions`, local `i_version` will always be larger than the one from kclient.

IMO we can add a new `i_wr_version` in CInode, and it will only record the `Fw` caps dirty version. In MDS we will update the `i_wr_version` only when clients dirty and flush the `Fw` caps to MDS. Sounds okay ?

jtlayton commented:
The client and MDS code is confusing in this matter, but when I mentioned `i_version` I meant the `inode->i_version` in the kernel, not `ci->i_version`. The `inode->i_version` in the client is called the change_attr in the ceph side of the MDS code.

I don't think it gets bumped just because we projected the inode. It should only change when there are "real" changes.

Would that value work?

lxbsz commented:
> The client and MDS code is confusing in this matter, but when I mentioned `i_version` I meant the `inode->i_version` in the kernel, not `ci->i_version`. The `inode->i_version` in the client is called the `change_attr` in the ceph side of the MDS code.
> 
> I don't think it gets bumped just because we projected the inode. It should only change when there are "real" changes.
> 
> Would that value work?

Okay, I was trying the `ci->i_version`. Let me investigate and test this to see whether will it work for us.

lxbsz commented:
> The client and MDS code is confusing in this matter, but when I mentioned `i_version` I meant the `inode->i_version` in the kernel, not `ci->i_version`. The `inode->i_version` in the client is called the `change_attr` in the ceph side of the MDS code.
> 
> I don't think it gets bumped just because we projected the inode. It should only change when there are "real" changes.
> 
> Would that value work?

Checked the code the `inode->i_version` is what I meant about the new `i_wr_version`. And locally I tested it and it works. Will test it more tomorrow.

lxbsz commented:
Rebased it to the upstream and only changed the commit `mds: add truncate size handling support for fscrypt` by switching the object version to `change_attr` to gate the operation. At the same time I will post the kclient related patch today.

vshankar commented:
jenkins test make check

lxbsz commented:
@jtlayton 

The conflicts are in the commit `client: add support for fscrypt_auth and fscrypt_file fields`, please take a look. 
Thanks.

jtlayton reviewed (approved):
LGTM
